### PR TITLE
Check response status in `OpenSearchDistribution.request()`

### PR DESCRIPTION
**Pull request by carlcsaposs-canonical**
## Issue

By default, `requests` does not raise an exception if an HTTP request errors (i.e. response status code 400-599)

Currently, `request()` is expecting an exception to be raised if request errors

For example, this usage of `request()` expects an `OpenSearchHttpError` if the status code is 400—but an exception will not be raised if the status code is 400 https://github.com/canonical/opensearch-operator/blob/c0225203320a412c3492ac827ac47b3b5b763787/lib/charms/opensearch/v0/opensearch_locking.py#L53-L68

Also, currently, requests with status code 400-599 are never retried, regardless of the value of `retries` argument

## Solution

Calling `raise_for_status()` will raise a `requests.exceptions.HTTPError` exception if the response status code is 400-599

### CODE REVIEW

**Comment by Mehdi-Bendriss** [comment]
Thanks! I'm not entirely sure, but I think it would have had to do with the `resp_status_code: bool` argument which would allow the request to only return an HTTP status code (error or success) instead of a response body or raised Exception - so `raise_for_status` would break this part. 
I've been meaning to get around to simplifying the `request(..)` method, and I believe it's about time we do it, i'll give it a try in this PR in addition to your change

File: lib/charms/opensearch/v0/opensearch_distro.py [hunk 1]
```diff
@@ -210,53 +208,16 @@ def request(  # noqa
...
-                    s.auth = (
-                        "admin",
-                        self._charm.secrets.get(Scope.APP, ADMIN_PW),
-                    )
+            for attempt in Retrying(
+                stop=stop_after_attempt(retries),
+                wait=wait_fixed(1),
+                before_sleep=error_http_retry_log(logger, retries, method, url, payload),
+                reraise=True,
+            ):
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_distro.py:218]: TIL, thanks! Indeed, the exceptions of interest are requests.RequestException and urllib3.HTTPError

File: lib/charms/opensearch/v0/opensearch_distro.py [hunk 2]
```diff
@@ -274,30 +235,35 @@ def call(
...
 
-        if resp_status_code:
-            return resp.status_code
+        try:
+            resp = call(urls[0])
+            if resp_status_code:
+                return resp.status_code
+        except (RequestException, HTTPError) as e:
+            if isinstance(e, HTTPError) or not e.response:
+                raise OpenSearchHttpError(response_body=str(e))
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_distro.py]: This is expected to handle HttpErrors (non-`requests.RequestException` i.e: host not reachable) and when a response is not part of the exception 
what concern do you see?

[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_distro.py]: In that case i think it should be: `if isinstance(e, HTTPError) or getattr(e, "response", None) is None:`

**Review by Mehdi-Bendriss — DISMISSED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -738,7 +738,6 @@ def _start_opensearch(self, event: EventBase) -> None:  # noqa: C901
         except OpenSearchHttpError:
             self.peers_data.delete(Scope.UNIT, "starting")
             event.defer()
-            self._post_start_init(event)
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:741]: it is correct

**Comment by Mehdi-Bendriss** [comment]
I added a fix to handle remaining exceptions, as the only exception expected to be raised from the `request` method is OpenSearchHttpError (this should fix the large-deployment integration test that failed)

**Comment by Mehdi-Bendriss** [comment]
Yes - I saw the error  - this is due to the certificates not being set on disk *yet* - the client making a request should not know / worry about any of this. 
Value errors / wrong arguments are 1 thing, we don't catch those. 
In general, catching all exception is bad - in general. Now the idea of a `request` function is to unify the handling of HTTP requests / responses / errors as far as inputs are correct - so the client should trust that only 1 exception is raised from it as it only cares about making a request, knowing it succeeded or if it failed.

**Comment by Mehdi-Bendriss** [comment]
> For the specific error in the integration test, my question was whether that's a valid input—is making a request without TLS configured a valid input?

It is not - but from the client's (charm) perspective, the cause of the failure of both "cert missing" / "post request" error is irrelevant. All it is interested in is: success / failure.
Keep in mind that we're not creating an http client lib, the only client of this method is the charm - and it's only tailored to its needs.

> regarding except Exception, IMO, a different exception (not just the 1 exception) should be raised if the request isn't even attempted

Good point, the thing is for the client the handling will be the same for both exceptions at the end. I suggest we add this later when/if we come across a case where the charm finds the need to differentiate between "any error" and "only pre/post-request errors"

What do you think?

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by phvalguima — APPROVED** [review]

**Comment by juditnovak** [comment]
@carlcsaposs-canonical @Mehdi-Bendriss 

Late to the party... I'd only have one question: why no tests at all...? There should be a way to verify this amount of changes (other than "it doesn't break anything" i.e. old tests pass), no...?

**Pull request merged**